### PR TITLE
Multi node test hardening for Sharding #27749

### PR DESCRIPTION
**Pull request by helena**
#27749

Tested in CI with the same Jenkins configs as our multi-node-nightly and multi-jvm.

Some interesting insights came from this exercise. For example, we should consider that the setup and initializations we do show users how to do things as well as reflect how they are initializing, configuring, updating on deploys for fewer surprises.
* While some specific tests do particular steps to initialize and setup, most use the same, hence the consolidation
* One base test to change vs all tests with a config or other internal change (saving time, more peace of mind on regression. etc.)

### CODE REVIEW

File: akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/MultiNodeClusterShardingSpec.scala [hunk 1]
```diff
@@ -72,67 +105,112 @@ abstract class MultiNodeClusterShardingSpec(val config: MultiNodeClusterSharding
...
+            // let's pick one or two vs have 3
+            // this was here originally and used in one or two tests
+            cluster.state.isMemberUp(node(from).address)
+            // these two were in a few tests
+            cluster.state.members.unsorted.map(_.uniqueAddress) should contain(cluster.selfUniqueAddress)
+            cluster.state.members.unsorted.map(_.status) shouldEqual Set(MemberStatus.Up)
+          }
+        }
       }
+      onJoinedRunOnFrom
```
Comment (patriknw) [inline akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/MultiNodeClusterShardingSpec.scala]: would it be good with a barrier before this `onJoinedRunOnFrom`?

I'd prefer a more explicit function `() => Unit`

File: akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/MultiNodeClusterShardingSpec.scala [hunk 2]
```diff
@@ -72,67 +105,112 @@ abstract class MultiNodeClusterShardingSpec(val config: MultiNodeClusterSharding
...
+  protected def join(from: RoleName, to: RoleName, onJoinedRunOnFrom: => Unit = Unit, assert: Boolean = true): Unit = {
     runOn(from) {
-      Cluster(system).join(node(to).address)
-      awaitAssert {
-        Cluster(system).state.isMemberUp(node(from).address)
+      cluster.join(node(to).address)
+      if (assert) {
+        within(20.seconds) {
+          awaitAssert {
+            // let's pick one or two vs have 3
```
Comment (patriknw) [inline akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/MultiNodeClusterShardingSpec.scala]: I don't understand that comment

**Review by patriknw — COMMENTED** [review]
Quick scan. Looking very nice. I trust that the others review this in more detail next week.

File: akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/MultiNodeClusterShardingSpec.scala (hunk 1 above)
Comment (helena) [reply akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/MultiNodeClusterShardingSpec.scala]: Possibly, though none of the tests were doing that.
The patterns were:
* join, barrier
* join, assert, barrier
* join, startSharding, barrier
* join, startSharding, assert, barrier

Comment (patriknw) [reply akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/MultiNodeClusterShardingSpec.scala]: ok, for startSharding it shouldn't matter

File: akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/MultiNodeClusterShardingSpec.scala (hunk 2 above)
Comment (helena) [reply akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/MultiNodeClusterShardingSpec.scala]: The comment is that some tests use this one assertion, some the other two, we should consolidate and pick the most appropriate if possible.

File: akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/MultiNodeClusterShardingSpec.scala (hunk 1 above)
Comment (helena) [reply akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/MultiNodeClusterShardingSpec.scala]: `() => Unit` added

File: akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/MultiNodeClusterShardingSpec.scala (hunk 2 above)
Comment (chbatey) [reply akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/MultiNodeClusterShardingSpec.scala]: 👍 i think the first is fine, no need to check all the others are up

**Review by chbatey — COMMENTED** [review]
Looks like a good cleanup

File: akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/ClusterShardingGracefulShutdownSpec.scala [hunk 3]
```diff
@@ -87,76 +35,41 @@ class PersistentClusterShardingGracefulShutdownMultiJvmNode2 extends PersistentC
...
-      }
-    }
-
     "start some shards in both regions" in within(30.seconds) {
-      join(first, first)
-      join(second, first)
+      startPersistenceIfNotDdataMode(startOn = first, setStoreOn = Seq(first, second))
+
+      join(first, first, typeName = "Entity")
+      join(second, first, typeName = "Entity")
```
Comment (johanandren) [inline akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/ClusterShardingGracefulShutdownSpec.scala]: Perhaps extract this typeName to a field and reference that in the (3?) places it is now hardcoded.

File: akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/ClusterShardingCustomShardAllocationSpec.scala [hunk 4]
```diff
@@ -81,37 +61,23 @@ object ClusterShardingCustomShardAllocationSpec {
 
 }
 
-abstract class ClusterShardingCustomShardAllocationSpecConfig(val mode: String) extends MultiNodeConfig {
+abstract class ClusterShardingCustomShardAllocationSpecConfig(mode: String)
+    extends MultiNodeClusterShardingConfig(
+      mode,
+      additionalConfig = ConfigFactory.parseString(s"""
```
Comment (johanandren) [inline akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/ClusterShardingCustomShardAllocationSpec.scala]: I think `additionalConfig` is always a string parsed to config, could perhaps be nudge cleaner if it was accepted as a string and parsed in `MultiNodeClusterShardingConfig` instead?

File: akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/ClusterShardingLeavingSpec.scala [hunk 5]
```diff
@@ -106,79 +66,32 @@ class DDataClusterShardingLeavingMultiJvmNode2 extends DDataClusterShardingLeavi
...
-      join(first, first)
-      join(second, first)
-      join(third, first)
-      join(fourth, first)
+      startPersistenceIfNotDdataMode(startOn = first, setStoreOn = roles)
+
+      join(first, first, onJoinedRunOnFrom = startSharding())
+      join(second, first, onJoinedRunOnFrom = startSharding())
+      join(third, first, onJoinedRunOnFrom = startSharding())
+      join(fourth, first, onJoinedRunOnFrom = startSharding())
```
Comment (johanandren) [inline akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/ClusterShardingLeavingSpec.scala]: Could just be `startSharding()` after this set of joins instead?

File: akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/MultiNodeClusterShardingSpec.scala [hunk 6]
```diff
@@ -72,67 +105,119 @@ abstract class MultiNodeClusterShardingSpec(val config: MultiNodeClusterSharding
...
+   *                      and similar assertions are not run. This allows tests that were
+   *                      not doing assertions (e.g. ClusterShardingMinMembersSpec) or
+   *                      doing them after `onJoinedRunOnFrom` more flexibility.
+   *                      Defaults to true, running member status checks.
+   */
+  protected def join(
+      from: RoleName,
+      to: RoleName,
+      onJoinedRunOnFrom: => Unit = (),
+      assertEnabled: Boolean = true,
```
Comment (johanandren) [inline akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/MultiNodeClusterShardingSpec.scala]: `assertNodeUp` or something else saying what it is actually asserting?

**Review by johanandren — APPROVED** [review]
LGTM with some opinions here and there.

File: akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/ClusterShardingGracefulShutdownSpec.scala (hunk 3 above)
Comment (helena) [reply akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/ClusterShardingGracefulShutdownSpec.scala]: Done

File: akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/ClusterShardingLeavingSpec.scala (hunk 5 above)
Comment (helena) [reply akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/ClusterShardingLeavingSpec.scala]: I wanted to keep the tests as much the same as they were.
This was calling `join` then `startSharding` in this order of operations as it still is now, with `startSharding` on the join `from` node.

File: akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/ClusterShardingQueriesSpec.scala [hunk 7]
```diff
@@ -111,11 +68,22 @@ abstract class ClusterShardingQueriesSpec
       awaitClusterUp(controller, first, second, third)
 
       runOn(controller) {
-        startProxy()
+        startProxy(
```
[the remark this answers is not in the document]
Comment (helena) [reply akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/ClusterShardingQueriesSpec.scala]: Formatter, if that's what you mean?

File: akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/MultiNodeClusterShardingSpec.scala (hunk 6 above)
Comment (helena) [reply akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/MultiNodeClusterShardingSpec.scala]: Done

File: akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/MultiNodeClusterShardingSpec.scala (hunk 2 above)
Comment (helena) [reply akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/MultiNodeClusterShardingSpec.scala]: LGTM thanks @chbatey

File: akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/ClusterShardingCustomShardAllocationSpec.scala [hunk 8]
```diff
@@ -126,62 +92,35 @@ class PersistentClusterShardingCustomShardAllocationMultiJvmNode2
...
-  def startSharding(): Unit = {
-    ClusterSharding(system).start(
-      typeName = "Entity",
-      entityProps = Props[Entity],
-      settings = ClusterShardingSettings(system),
-      extractEntityId = extractEntityId,
-      extractShardId = extractShardId,
-      allocationStrategy = TestAllocationStrategy(allocator),
-      handOffStopMessage = PoisonPill)
+    join(
```
[the remark this answers is not in the document]
Comment (helena) [reply akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/ClusterShardingCustomShardAllocationSpec.scala]: It's a heat-seeking missile yes

File: akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/ClusterShardingCustomShardAllocationSpec.scala (hunk 4 above)
Comment (helena) [reply akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/ClusterShardingCustomShardAllocationSpec.scala]: Good, done.

File: akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/ClusterShardingLeavingSpec.scala (hunk 5 above)
Comment (johanandren) [reply akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/ClusterShardingLeavingSpec.scala]: Alright, no strong feelings about it more than that i stumbled on the repetition. Ok!

**Review by johanandren — APPROVED** [review]
LGTM!

**Review by chbatey — APPROVED** [review]
LGTM

**Comment by helena** [comment]
Github says, "Investigating - We are investigating reports of increased API errors and webhook latency. Feb 27, 14:31 UTC” so I will merge because I've signed a long time ago
Screen Shot 2020-02-27 at 12 04 00 PM

**Pull request merged**